File: iOS/DateMate/UI/SignIn/SignInViewModel.swift
//
//  SignInViewModel.swift
//  DateMate
//
//  Created by 홍희표 on 2021/11/28.
//

import Foundation
import FirebaseAuth
import Combine

class SignInViewModel: ObservableObject {
    private let userRepository: UserRepository
    
    @Published var uiState = SignInUiState()
    
    private var cancellables = Set<AnyCancellable>()
    
    private func canSubmit(email: String, password: String) -> Bool {
        return !email.isEmpty &&
        validateEmail(email) == nil &&
        !password.isEmpty &&
        validatePassword(password) == nil
    }

    private func isEmailValid(_ email: String) -> Bool {
        let regex = #"^(.+)@(.+)$"#
        return email.range(of: regex, options: .regularExpression) != nil
    }
    
    private func validateEmail(_ value: String) -> String? {
        if value.isEmpty || isEmailValid(value) {
            return nil
        } else {
            return "Invalid email: \(value)"
        }
    }

    private func isPasswordValid(_ password: String) -> Bool {
        return password.count > 3
    }

    private func validatePassword(_ value: String) -> String? {
        if value.isEmpty || isPasswordValid(value) {
            return nil
        } else {
            return "Invalid password"
        }
    }
    
    func onEmailChanged(_ value: String) {
        let password = uiState.password
        uiState.email = value
        uiState.emailError = validateEmail(value)
        uiState.isSignInEnabled = canSubmit(email: value, password: password)
    }
    
    func onPasswordChanged(_ value: String) {
        let email = uiState.email
        uiState.password = value
        uiState.passwordError = validatePassword(value)
        uiState.isSignInEnabled = canSubmit(email: email, password: value)
    }
    
    func signIn(email: String, password: String) {
        userRepository.getSignInResultStream(email: email, password: password)
            .receive(on: DispatchQueue.main)
            .sink { result in
                switch result.state {
                case .Loading:
                    self.uiState.isLoading = true
                case .Success:
                    self.uiState.isLoading = false
                case .Error:
                    self.uiState.isLoading = false
                }
            }
            .store(in: &cancellables)
    }
    
    init(_ userRepository: UserRepository) {
        self.userRepository = userRepository
    }
}

File: Android/app/src/main/java/com/hhp227/datemate/ui/signin/SignInViewModel.kt
package com.hhp227.datemate.ui.signin

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hhp227.datemate.data.repository.UserRepository
import com.hhp227.datemate.common.Resource
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.regex.Pattern

class SignInViewModel internal constructor(
    private val userRepository: UserRepository
) : ViewModel() {
    private val _uiState = MutableStateFlow(SignInUiState())
    val uiState: StateFlow<SignInUiState> = _uiState

    private fun canSubmit(email: String, password: String): Boolean {
        return !email.isEmpty() && validateEmail(email) == null && !password.isEmpty() && validatePassword(password) == null
    }

    private fun isEmailValid(email: String): Boolean {
        return Pattern.matches(EMAIL_VALIDATION_REGEX, email)
    }

    private fun validateEmail(value: String): String? {
        return if (value.isEmpty() || isEmailValid(value)) null else "Invalid email: $value"
    }

    private fun isPasswordValid(password: String): Boolean {
        return password.length > 3
    }

    private fun validatePassword(value: String): String? {
        return if (value.isEmpty() || isPasswordValid(value)) null else "Invalid password"
    }

    fun onEmailChanged(value: String) {
        _uiState.update {
            it.copy(
                email = value,
                emailError = validateEmail(value),
                isSignInEnabled = canSubmit(value, it.password)
            )
        }
    }

    fun onPasswordChanged(value: String) {
        _uiState.update {
            it.copy(
                password = value,
                passwordError = validatePassword(value),
                isSignInEnabled = canSubmit(it.email, value)
            )
        }
    }

    fun signIn(email: String, password: String) {
        viewModelScope.launch {
            userRepository.getSignInResultStream(email, password)
                .collectLatest { result ->
                    when (result) {
                        is Resource.Error<*> -> Unit
                        is Resource.Loading<*> -> _uiState.update { it.copy(isLoading = true) }
                        is Resource.Success<*> -> Unit
                    }
                }
        }
    }

    companion object {
        private const val EMAIL_VALIDATION_REGEX = "^(.+)@(.+)\$"
    }
}